e2e/e2e_test.go

```diff
@@ -0,0 +1,301 @@
+// +build integration
+
+/*
+Copyright 2017 DigitalOcean
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+	"path"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/davecgh/go-spew/spew"
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+const (
+	numWantNodes          = 2
+	doLabel               = "beta.kubernetes.io/instance-type"
+	kopsEnvVarClusterName = "KOPS_CLUSTER_NAME"
+	kopsEnvVarStateStore  = "KOPS_STATE_STORE"
+)
+
+// TestE2E verifies that the node and service controller work as intended for
+// all supported Kubernetes versions; that is, we expect nodes to become ready
+// and requests to be routed through a DO-provisioned load balancer.
+// The test creates various components and makes sure they get deleted prior to
+// (to clean up any previous left-overs) and after testing.
+func TestE2E(t *testing.T) {
+	var missingEnvs []string
+	for _, env := range []string{kopsEnvVarClusterName, kopsEnvVarStateStore} {
+		if _, ok := os.LookupEnv(env); !ok {
+			missingEnvs = append(missingEnvs, env)
+		}
+	}
+	if len(missingEnvs) > 0 {
+		t.Fatalf("missing required environment variable(s): %s", missingEnvs)
+	}
+
+	s3Cl, err := createS3Client()
+	if err != nil {
+		t.Fatalf("failed to create S3 client: %s", err)
+	}
+
+	tests := []struct {
+		desc    string
+		kubeVer string
+	}{
+		{
+			desc:    "latest release",
+			kubeVer: "1.12.0",
+		},
+		{
+			desc:    "previous release",
+			kubeVer: "1.11.2",
+		},
+		{
+			desc:    "previous previous release",
+			kubeVer: "1.10.6",
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.desc, func(t *testing.T) {
+			l := log.New(os.Stdout, fmt.Sprintf("[%s] ", t.Name()), 0)
+			dnsName := toDNSName(t.Name())
+
+			wd, err := os.Getwd()
+			if err != nil {
+				t.Fatalf("failed to get working directory: %s", err)
+			}
+			kubeConfFile := path.Join(wd, "kubeconfig-e2e."+dnsName)
+
+			// Delete old kubeconfig
+			if err := os.Remove(kubeConfFile); err != nil && !os.IsNotExist(err) {
+				t.Fatalf("failed to delete kubeconfig %q: %s", kubeConfFile, err)
+			}
+
+			// Create space.
```

Comment:
I would prefer to keep all the cluster setup steps to be done via bash or at least in a separate go function to separate where we do setup vs testing. what do you think?

Reply:
Also, might be easier to just assume a bucket is created and always use that bucket, kops should handle subpath creation from there.

Reply:
ah, seems like we did this here so we can set the version on the test table?

Reply:
>ah, seems like we did this here so we can set the version on the test table?

Yes, exactly: we couldn't do this in bash alone as we need to parameterize kops for at least the Kubernetes version and the state store URL which are currently driven by the Go test.
One alternative approach might be to let (a smarter version of) the bash script be in charge and call specific (sub-)tests with the right parameters. Personally though, I'd rather have more logic in Go than in bash as it's easier to understand and modify that way.
Does that address your comments?

I'll follow your advice for sure to group the test setup and test execution into separate functions for better readability. 👍

Reply:
I should also mention that I started off my PR with a static bucket but moved away from it for two reasons:

1. I ran into at least one scenario where not having a clean state to start from caused trouble: I once tried to change the number of worker nodes, which from kops' perspective seems to be a change to an existing cluster. Unfortunately, kops does not support downsizing an k8s cluster in DO, so my test failed even though it was entirely unrelated to the system-under-test. Nobody will likely do this for the test once it's solid, but it showed me that relying on in-place adjustments is a different mode of operation, so I decided to get rid of it altogether by having separate clusters per test (and cleaning up accordingly).
1. Dynamic buckets that can be removed after test completion means no further cost for the user that runs an end-to-end test and forgets to clean up manually.

If we want a static bucket for simplicity reasons, I'm okay to make the change. Just wanted to make sure the trade-offs are clear before making a decision.

Reply:
Just pushed a few minor improvements / fixes along with a change that moves the setup part into a dedicated function.

One extra complication with the latter is that we now need the caller to (defer-)invoke the cleanup functions defined during the setup (e.g., store state removal). It takes a bit of iterating and error aggregation to make it happen but I personally think it's worth it, and the pattern is quite common.

Happy to hear your thoughts, and ready to make any further adjustments we may find relevant. :)